sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkGetFunctionsOperation.scala

```diff
@@ -0,0 +1,118 @@
+import org.apache.spark.sql.SQLContext
+import org.apache.spark.util.{Utils => SparkUtils}
+
+/**
+ * Spark's own GetTablesOperation
+ *
+ * @param sqlContext SQLContext to use
+ * @param parentSession a HiveSession from SessionManager
+ * @param catalogName catalog name. null if not applicable
+ * @param schemaName database name, null or a concrete database name
+ * @param functionName function name pattern
+ */
+private[hive] class SparkGetFunctionsOperation(
+    sqlContext: SQLContext,
+    parentSession: HiveSession,
+    catalogName: String,
+    schemaName: String,
+    functionName: String)
+  extends GetFunctionsOperation(parentSession, catalogName, schemaName, functionName)
+    with Logging {
+
+  override def runInternal(): Unit = {
+    val statementId = UUID.randomUUID().toString
+    // Do not change cmdStr. It's used for Hive auditing and authorization.
+    val cmdStr = s"catalog : $catalogName, schemaPattern : $schemaName"
+    val logMsg = s"Listing functions '$cmdStr'"
+    logInfo(s"$logMsg with $statementId")
+    setState(OperationState.RUNNING)
+    // Always use the latest class loader provided by executionHive's state.
+    val executionHiveClassLoader = sqlContext.sharedState.jarClassLoader
+    Thread.currentThread().setContextClassLoader(executionHiveClassLoader)
+
+    val catalog = sqlContext.sessionState.catalog
+
+    if (isAuthV2Enabled) {
+      // get databases for schema pattern
+      val schemaPattern = convertSchemaPattern(schemaName)
+      var matchingDbs: Seq[String] = null
+      try {
+        matchingDbs = catalog.listDatabases(schemaPattern)
+      } catch {
+        case e: TException =>
+          setState(OperationState.ERROR)
+          HiveThriftServer2.listener.onStatementError(
+            statementId, e.getMessage, SparkUtils.exceptionString(e))
+          throw new HiveSQLException(e)
+      }
+      // authorize this call on the schema objects
+      val privObjs =
+        HivePrivilegeObjectUtils.getHivePrivDbObjects(seqAsJavaListConverter(matchingDbs).asJava)
+      authorizeMetaGets(HiveOperationType.GET_FUNCTIONS, privObjs, cmdStr)
+    }
+
+    HiveThriftServer2.listener.onStatementStart(
+      statementId,
+      parentSession.getSessionHandle.getSessionId.toString,
+      logMsg,
+      statementId,
+      parentSession.getUsername)
+
+    try {
+      val functionPattern = CLIServiceUtils.patternToRegex(functionName)
+      if ((null == catalogName || "".equals(catalogName))
+        && (null == schemaName || "".equals(schemaName))) {
+        catalog.listFunctions(catalog.getCurrentDatabase, functionPattern).foreach {
+          case (functionIdentifier, _) =>
+            val rowData = Array[AnyRef](
+              null, // FUNCTION_CAT
+              null, // FUNCTION_SCHEM
+              functionIdentifier.funcName, // FUNCTION_NAME
+              "", // REMARKS
+              DatabaseMetaData.functionResultUnknown.asInstanceOf[AnyRef], // FUNCTION_TYPE
```

Comment:
We do not support `FUNCTION_TYPE` now. Set it to Unknown:
```java
    // java.sql.DatabaseMetaData

    /**
     * Indicates that it is not known whether the function returns
     * a result or a table.
     * <P>
     * A possible value for column <code>FUNCTION_TYPE</code> in the
     * <code>ResultSet</code> object returned by the method
     * <code>getFunctions</code>.
     * @since 1.6
     */
    int functionResultUnknown   = 0;

    /**
     * Indicates that the function  does not return a table.
     * <P>
     * A possible value for column <code>FUNCTION_TYPE</code> in the
     * <code>ResultSet</code> object returned by the method
     * <code>getFunctions</code>.
     * @since 1.6
     */
    int functionNoTable         = 1;

    /**
     * Indicates that the function  returns a table.
     * <P>
     * A possible value for column <code>FUNCTION_TYPE</code> in the
     * <code>ResultSet</code> object returned by the method
     * <code>getFunctions</code>.
     * @since 1.6
     */
    int functionReturnsTable    = 2;
```